File: tools/trainexport/GoldenGuard.cs
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace TrainExport;

/// <summary>
/// Refuses to let anything from the golden benchmark into a training export.
/// </summary>
/// <remarks>
/// <para>Golden data may be read to understand what is failing. It may not be trained on.
/// Every published quality number in <c>docs/</c> is a statement about those exact 841
/// items, and a model that has seen them turns the whole series of before/after tables into
/// a measurement of memorisation — silently, and irreversibly, because there is no way to
/// un-train a model or to tell after the fact which items it had seen.</para>
///
/// <para>Blocking by id alone would be theatre: the ids never leave the benchmark file, and
/// nothing would stop a pipeline that copied the sentences. The guard is therefore on
/// <em>content</em> — a normalised hash of every golden sentence, in both its broken and its
/// corrected form — with ids kept as a second, cheaper check. Normalisation folds case,
/// collapses whitespace and drops punctuation, so re-punctuating or re-casing a golden
/// sentence does not smuggle it through.</para>
///
/// <para>The exporter counts what it blocks and writes that count into the manifest. Zero is
/// the expected number and a non-zero one is a finding, not a nuisance: it means a source
/// corpus overlaps the benchmark and the overlap needs explaining.</para>
/// </remarks>
internal sealed class GoldenGuard
{
    private readonly HashSet<string> _sentenceHashes = new(StringComparer.Ordinal);
    private readonly HashSet<string> _ids = new(StringComparer.Ordinal);

    public int BlockedCount { get; private set; }

    public int GoldenSentenceCount => _sentenceHashes.Count;

    public int GoldenItemCount => _ids.Count;

    public static GoldenGuard FromCorpus(string goldenPath)
    {
        var guard = new GoldenGuard();
        foreach (var line in File.ReadLines(goldenPath))
        {
            if (string.IsNullOrWhiteSpace(line)) continue;
            using var document = JsonDocument.Parse(line);
            var root = document.RootElement;

            if (root.TryGetProperty("id", out var id) && id.GetString() is { Length: > 0 } idText)
            {
                guard._ids.Add(idText);
            }

            // Both forms. The corrected sentence is what a training example would most
            // naturally contain, and it is the one an exporter is most likely to reach for.
            foreach (var field in new[] { "source", "target" })
            {
                if (root.TryGetProperty(field, out var value) && value.GetString() is { Length: > 0 } text)
                {
                    guard._sentenceHashes.Add(Fingerprint(text));
                }
            }
        }

        return guard;
    }

    /// <summary>True when this example must not be exported.</summary>
    public bool IsGolden(string sentence, string? id = null)
    {
        var blocked = (id is not null && _ids.Contains(id)) || _sentenceHashes.Contains(Fingerprint(sentence));
        if (blocked)
        {
            BlockedCount++;
        }

        return blocked;
    }

    /// <summary>
    /// A content fingerprint that survives re-casing, re-spacing and re-punctuation.
    /// </summary>
    /// <remarks>
    /// Punctuation is dropped deliberately, even though one of the exported tasks is about
    /// punctuation: the question here is "is this the same sentence", and a golden item with
    /// its commas moved is still the golden item.
    /// </remarks>
    public static string Fingerprint(string sentence)
    {
        var builder = new StringBuilder(sentence.Length);
        var lastWasSpace = true;
        foreach (var ch in sentence)
        {
            if (char.IsLetterOrDigit(ch))
            {
                builder.Append(char.ToLowerInvariant(ch) == 'ё' ? 'е' : char.ToLowerInvariant(ch));
                lastWasSpace = false;
            }
            else if (!lastWasSpace)
            {
                builder.Append(' ');
                lastWasSpace = true;
            }
        }

        var normalised = builder.ToString().Trim();
        return Convert.ToHexStringLower(SHA256.HashData(Encoding.UTF8.GetBytes(normalised)));
    }
}

File: ai/scripts/golden_guard.py
"""Refuses to let a training run start on data that overlaps the golden benchmark.

The exporter already blocks golden content on the way out (`tools/trainexport/GoldenGuard.cs`),
and `GoldenDatasetIntegrityTests` checks the written artefacts. This is the third check and it
sits at the last possible moment: immediately before a model sees an example.

That placement is the point. The first two guard the pipeline that produced today's files;
this one guards the file actually being read, whatever produced it — a hand-edited jsonl, a
copy from another machine, a future exporter with a new bug. A model that has seen the
benchmark cannot be un-trained, and there is no way to tell after the fact which items it saw,
so the failure has to be loud and it has to be fatal. `verify_or_die` raises SystemExit; there
is deliberately no flag to downgrade it to a warning.

The fingerprint matches the C# implementation exactly: letters and digits only, lowercased,
ё folded to е, everything else collapsed to a single space. Re-punctuating or re-casing a
golden sentence therefore does not smuggle it past.
"""

from __future__ import annotations

import hashlib
import io
import json
import os


def fingerprint(sentence: str) -> str:
    """Content hash that survives re-casing, re-spacing and re-punctuation."""
    out: list[str] = []
    last_was_space = True
    for ch in sentence:
        if ch.isalpha() or ch.isdigit():
            lowered = ch.lower()
            out.append("е" if lowered == "ё" else lowered)
            last_was_space = False
        elif not last_was_space:
            out.append(" ")
            last_was_space = True
    return hashlib.sha256("".join(out).strip().encode("utf-8")).hexdigest()


def load_golden(path: str) -> tuple[set[str], set[str]]:
    """Returns (content fingerprints, ids) for every golden sentence, both forms."""
    fingerprints: set[str] = set()
    ids: set[str] = set()
    with io.open(path, encoding="utf-8") as handle:
        for line in handle:
            line = line.strip()
            if not line:
                continue
            record = json.loads(line)
            if record.get("id"):
                ids.add(record["id"])
            for field in ("source", "target"):
                if record.get(field):
                    fingerprints.add(fingerprint(record[field]))
    return fingerprints, ids


def verify_or_die(
    records: list[dict],
    golden_path: str,
    text_field: str = "sentence",
    what: str = "training data",
) -> dict:
    """Aborts the process if any record overlaps golden. Returns guard stats when clean."""
    if not os.path.exists(golden_path):
        raise SystemExit(
            f"golden corpus not found at {golden_path}; refusing to train without a leakage guard"
        )

    fingerprints, ids = load_golden(golden_path)
    leaks: list[str] = []
    for record in records:
        text = record.get(text_field, "")
        if fingerprint(text) in fingerprints or record.get("id") in ids:
            leaks.append(record.get("id", "<no id>"))
            if len(leaks) >= 10:
                break

    if leaks:
        raise SystemExit(
            f"GOLDEN LEAKAGE in {what}: {len(leaks)}+ records overlap the frozen benchmark "
            f"by normalised content fingerprint or id. First offenders: {leaks}. "
            "Training is aborted. The golden set is evaluation-only; every published quality "
            "number is a statement about those exact items."
        )

    return {
        "goldenPath": golden_path,
        "goldenFingerprints": len(fingerprints),
        "goldenIds": len(ids),
        "recordsChecked": len(records),
        "leaks": 0,
    }
